### PR TITLE
- Reactivate Generator function in Unity 2018.4.

### PR DESCRIPTION
**Pull request by fakegood**
Since Generator is one of the key features of Unity Atoms, this PR is to re-activate this feature. And since 2018.4 is LTS version I think it should be supported as well.

### CODE REVIEW

File: Packages/Core/Editor/Editors/AtomEventEditor.cs [hunk 1]
```diff
@@ -1,6 +1,10 @@
-#if UNITY_2019_1_OR_NEWER
+#if UNITY_2018_4_OR_NEWER
 using UnityEngine;
+#if UNITY_2018_4_OR_NEWER
```
Comment (AdamRamberg) [inline Packages/Core/Editor/Editors/AtomEventEditor.cs]: The UNITY_2019_1_OR_NEWER condition should be before UNITY_2018_4_OR_NEWER, the second if-statement will never be used when running Unity 2019 since the UNITY_2018_4_OR_NEWER flag will also be true in this case.

File: Packages/Core/Editor/Editors/AtomEventEditor.cs [hunk 2]
```diff
@@ -12,6 +16,9 @@ namespace UnityAtoms.Editor
     public abstract class AtomEventEditor<T, E> : UnityEditor.Editor
         where E : AtomEvent<T>
     {
+#if UNITY_2018_4_OR_NEWER
```
Comment (AdamRamberg) [inline Packages/Core/Editor/Editors/AtomEventEditor.cs]: Same as above. The UNITY_2019_1_OR_NEWER condition should be before UNITY_2018_4_OR_NEWER, the second if-statement will never be used when running Unity 2019 since the UNITY_2018_4_OR_NEWER flag will also be true in this case.

File: Examples/Packages/manifest.json [hunk 3]
```diff
@@ -9,15 +9,8 @@
     "com.mambojambostudios.unity-atoms-tags": "file:../../Packages/Tags",
     "com.mambojambostudios.unity-atoms-ui": "file:../../Packages/UI",
     "com.neuecc.unirx": "https://github.com/starikcetin/UniRx.git#7.1.0-upm",
-    "com.unity.2d.sprite": "1.0.0",
```
Comment (AdamRamberg) [inline Examples/Packages/manifest.json]: Even though we support 2018.4, the Example project itself should run Unity 2019, and some of these packages are needed in that version. Please revert this file.

File: Packages/Core/Editor/Generator/Generator.cs [hunk 4]
```diff
@@ -58,7 +58,11 @@ public void Generate(string type, string baseWritePath, bool isEquatable, List<A
             var templateSearchPath = Runtime.IsUnityAtomsRepo ?
                 Directory.GetParent(baseWritePath).FullName : // "Packages"
                 Directory.GetParent(Application.dataPath).FullName;
+#if UNITY_2018_4_OR_NEWER
```
Comment (AdamRamberg) [inline Packages/Core/Editor/Generator/Generator.cs]: Same as above. The UNITY_2019_1_OR_NEWER condition should be before UNITY_2018_4_OR_NEWER, the second if-statement will never be used when running Unity 2019 since the UNITY_2018_4_OR_NEWER flag will also be true in this case.

File: Packages/Core/Editor/Generator/Generator.cs [hunk 5]
```diff
@@ -70,20 +74,25 @@ public void Generate(string type, string baseWritePath, bool isEquatable, List<A
 
             foreach (var templatePath in templatePaths)
             {
-                var templateNameStartIndex = templatePath.LastIndexOf(Path.DirectorySeparatorChar) + 1;
+#if UNITY_2018_4_OR_NEWER
```
Comment (AdamRamberg) [inline Packages/Core/Editor/Generator/Generator.cs]: Same as above + add UNITY_2019_1_OR_NEWER condition. The UNITY_2019_1_OR_NEWER condition should be before UNITY_2018_4_OR_NEWER, the second if-statement will never be used when running Unity 2019 since the UNITY_2018_4_OR_NEWER flag will also be true in this case.

File: Packages/Core/Editor/Generator/GeneratorEditor.cs [hunk 6]
```diff
@@ -1,10 +1,16 @@
-#if UNITY_2019_1_OR_NEWER
+#if UNITY_2018_4_OR_NEWER
 using System;
 using System.Linq;
 using System.Collections.Generic;
 using UnityEditor;
 using UnityEngine;
+#if UNITY_2018_4_OR_NEWER
```
Comment (AdamRamberg) [inline Packages/Core/Editor/Generator/GeneratorEditor.cs]: Same as above. The UNITY_2019_1_OR_NEWER condition should be before UNITY_2018_4_OR_NEWER, the second if-statement will never be used when running Unity 2019 since the UNITY_2018_4_OR_NEWER flag will also be true in this case.

File: Packages/Core/Editor/Generator/GeneratorEditor.cs [hunk 7]
```diff
@@ -147,7 +153,11 @@ private void OnEnable()
             };
             generator = generator == null ? new Generator() : generator;
 
+#if UNITY_2018_4_OR_NEWER
```
Comment (AdamRamberg) [inline Packages/Core/Editor/Generator/GeneratorEditor.cs]: Same as above. The UNITY_2019_1_OR_NEWER condition should be before UNITY_2018_4_OR_NEWER, the second if-statement will never be used when running Unity 2019 since the UNITY_2018_4_OR_NEWER flag will also be true in this case.

File: Examples/Packages/manifest.json [hunk 8]
```diff
@@ -59,4 +59,4 @@
       "revision": "7.1.0-upm"
     }
   }
-}
+}
```
Comment (AdamRamberg) [inline Examples/Packages/manifest.json]: Please revert this change

File: Packages/Core/Editor/Generator/Generator.cs [hunk 9]
```diff
@@ -58,7 +58,11 @@ public void Generate(string type, string baseWritePath, bool isEquatable, List<A
             var templateSearchPath = Runtime.IsUnityAtomsRepo ?
                 Directory.GetParent(baseWritePath).FullName : // "Packages"
                 Directory.GetParent(Application.dataPath).FullName;
+#if UNITY_2019_1_OR_NEWER
```
Comment (AdamRamberg) [inline Packages/Core/Editor/Generator/Generator.cs]: This check is not necessary. You should always be able to do: 
`var templatePaths = Directory.GetFiles(templateSearchPath, "UA_Template*.txt", SearchOption.AllDirectories);`

File: Packages/Core/Editor/Generator/Generator.cs [hunk 10]
```diff
@@ -70,20 +74,25 @@ public void Generate(string type, string baseWritePath, bool isEquatable, List<A
 
             foreach (var templatePath in templatePaths)
             {
-                var templateNameStartIndex = templatePath.LastIndexOf(Path.DirectorySeparatorChar) + 1;
+#if UNITY_2019_1_OR_NEWER
```
Comment (AdamRamberg) [inline Packages/Core/Editor/Generator/Generator.cs]: All the changes from this line and below is not necessary, it should work with just using `templatePath`.

File: Packages/Core/Editor/Generator/Generator.cs [hunk 11]
```diff
@@ -70,20 +74,25 @@ public void Generate(string type, string baseWritePath, bool isEquatable, List<A
...
+                var localTemplatePath = AssetDatabase.GUIDToAssetPath(templatePath);
+#endif
+                var templateNameStartIndex = localTemplatePath.LastIndexOf(Path.DirectorySeparatorChar) + 1;
                 var fileExtLength = 4;
-                var templateName = templatePath.Substring(templateNameStartIndex, templatePath.Length - templateNameStartIndex - fileExtLength);
+                var templateName = localTemplatePath.Substring(templateNameStartIndex, localTemplatePath.Length - templateNameStartIndex - fileExtLength);
                 var lastIndexOfDoubleUnderscore = templateName.LastIndexOf("__");
                 var atomType = templateName.Substring(lastIndexOfDoubleUnderscore + 2);
-                var capitalizedAtomType = Capitalize(atomType);
+                var capitalizedAtomType = Capitalize(atomType); Debug.Log(localTemplatePath + " : " + templateName);
```
Comment (AdamRamberg) [inline Packages/Core/Editor/Generator/Generator.cs]: Don't forget to remove the `Debug.Log`

**Comment by fakegood** [comment]
Updated based on comments.

Added path to Core Package because the template search path should not be based on a variable path input by user.

When Example project is opened, changed Generator path to "Assets/Atoms" will cause search path to search in "Assets/Atoms" instead of "../Packages/Core", which will not be able to search for any templates.

File: Packages/Core/Editor/Generator/Generator.cs [hunk 12]
```diff
@@ -56,8 +56,9 @@ public void Generate(string type, string baseWritePath, bool isEquatable, List<A
 
             // Recursively search for template files. TODO: Is there a better way to find and load templates?
             var templateSearchPath = Runtime.IsUnityAtomsRepo ?
-                Directory.GetParent(baseWritePath).FullName : // "Packages"
+                Directory.GetParent(Runtime.UnityAtomsCorePath).FullName : // "Packages"
```
Comment (AdamRamberg) [inline Packages/Core/Editor/Generator/Generator.cs]: I would prefer if you do it like this instead:
`Directory.GetParent(Application.dataPath).Parent.FullName`
In my mind it's better if we can compute the path at runtime instead of relying on constants.

File: Packages/Core/Runtime/Utils/Runtime.cs [hunk 13]
```diff
@@ -26,5 +26,14 @@ public static bool IsUnityAtomsRepo
         {
             get => System.Environment.CurrentDirectory.Contains("unity-atoms/Examples");
         }
+
+        /// <summary>
+        /// Returns the path to Unity Atoms Core.
+        /// Mainly used to search for Templates.
+        /// </summary>
+        public static string UnityAtomsCorePath
```
Comment (AdamRamberg) [inline Packages/Core/Runtime/Utils/Runtime.cs]: Please remove constant. See my first comment.

**Comment by fakegood** [comment]
Updated based on your comments. :)

File: Packages/Core/Editor/Drawers/VariableDrawer.cs [hunk 14]
```diff
@@ -14,15 +14,29 @@ public override void OnGUI(Rect position, SerializedProperty property, GUIConten
                 return;
             }
 
+            label = EditorGUI.BeginProperty(position, label, property);
```
Comment (AdamRamberg) [inline Packages/Core/Editor/Drawers/VariableDrawer.cs]: Sorry, but need you to revert these changes before I merge this PR (the complete file).

**Comment by fakegood** [comment]
I've updated the pull request.
